kubernetes/cache.go

```diff
@@ -110,26 +149,135 @@ func initControllers(clientset kube.Interface, refreshDuration time.Duration) ma
 	controllers["CronJob"] = sharedInformers.Batch().V1beta1().CronJobs().Informer()
 	controllers["Service"] = sharedInformers.Core().V1().Services().Informer()
 	controllers["Endpoints"] = sharedInformers.Core().V1().Endpoints().Informer()
+	// controllers["Namespaces"] = sharedInformers.Core().V1().Namespaces().Informer()
+
+	c.controllers[namespace] = controllers
+}
+
+func (c *controllerImpl) initIstioControllersNamespace(namespace string) {
+	// Caching only objects that are used in the validations
+
+	controllers := make(map[string]cache.SharedIndexInformer)
+	// Networking API
+	controllers[virtualServiceType] = createIstioIndexInformer(c.istioClient.istioNetworkingApi, virtualServices, c.refreshDuration, namespace)
+	controllers[destinationRuleType] = createIstioIndexInformer(c.istioClient.istioNetworkingApi, destinationRules, c.refreshDuration, namespace)
+	controllers[gatewayType] = createIstioIndexInformer(c.istioClient.istioNetworkingApi, gateways, c.refreshDuration, namespace)
+	controllers[serviceentryType] = createIstioIndexInformer(c.istioClient.istioNetworkingApi, serviceentries, c.refreshDuration, namespace)
+
+	// Authentication API
+	// controllers[meshPolicyType] = createIstioIndexInformer(c.istioClient.istioAuthenticationApi, meshPolicies, c.refreshDuration, namespace)
+	controllers[policyType] = createIstioIndexInformer(c.istioClient.istioAuthenticationApi, policies, c.refreshDuration, namespace)
+
+	// RBAC API
+	controllers[serviceroleType] = createIstioIndexInformer(c.istioClient.istioRbacApi, serviceroles, c.refreshDuration, namespace)
+	controllers[servicerolebindingType] = createIstioIndexInformer(c.istioClient.istioRbacApi, servicerolebindings, c.refreshDuration, namespace)
+	// controllers[clusterrbacconfigType] = createIstioIndexInformer(c.istioClient.istioRbacApi, clusterrbacconfigs, c.refreshDuration, namespace)
+
+	// Enable
+	c.istioControllers[namespace] = controllers
+}
+
+func createIstioIndexInformer(getter cache.Getter, resourceType string, refreshDuration time.Duration, namespace string) cache.SharedIndexInformer {
+	return cache.NewSharedIndexInformer(cache.NewListWatchFromClient(getter, resourceType, namespace, fields.Everything()),
+		&GenericIstioObject{},
+		refreshDuration,
+		cache.Indexers{},
+	)
+}
+
+func (c *controllerImpl) ProjectListener() cache.ResourceEventHandlerFuncs {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			project := obj.(*osproject_v1.Project)
+			namespace := project.GetObjectMeta().GetName()
+			c.controllerInitializer(namespace)
+		},
+		DeleteFunc: func(obj interface{}) {
+			project := obj.(*osproject_v1.Project)
+			namespace := project.GetObjectMeta().GetName()
+			c.stopNamespace(namespace)
+		},
+	}
+}
+
+func (c *controllerImpl) NamespaceListener() cache.ResourceEventHandlerFuncs {
```

Comment:
Interesting, I thought that the k8s/client-go should have already a Listener for Namespaces.

Reply:
Listener yes, but we want to attach the ResourceEventHandlerFunc to it so we can act on the events.